addons/mrp/models/mrp_bom.py

```diff
@@ -146,9 +176,12 @@ def check_cycle(v, visited, recStack, graph):
         V |= set([product.product_tmpl_id.id])
 
         bom_lines = [(bom_line, product, quantity, False) for bom_line in self.bom_line_ids]
+        bom_products = self.env['product.product'].browse()
```

Comment:
```suggestion
        bom_products = self.env['product.product']
```

By calling `browse()` AFAIK you're opening a new prefetch object, which probably should penalize performance here.